src/gui/properties/speedplotview.cpp

```diff
@@ -118,55 +113,93 @@ namespace
     }
 }
 
-SpeedPlotView::Averager::Averager(int divider, boost::circular_buffer<PointData> &sink)
-    : m_divider(divider)
-    , m_sink(sink)
-    , m_counter(0)
-    , m_accumulator {}
+SpeedPlotView::Averager::Averager(const milliseconds duration, const milliseconds resolution)
+    : m_resolution {resolution}
+    , m_maxDuration {duration}
+    , m_currentMaxDuration {m_maxDuration}
 {
+    m_timeSinceLastPush.start();
 }
 
-void SpeedPlotView::Averager::push(const PointData &pointData)
+bool SpeedPlotView::Averager::push(const PointData &pointData)
 {
     // Accumulator overflow will be hit in worst case on longest used averaging span,
     // defined by divider value. Maximum divider is DIVIDER_24HOUR = 144
     // Using int32 for accumulator we get overflow when transfer speed reaches 2^31/144 ~~ 14.2 MBytes/s.
     // With quint64 this speed limit is 2^64/144 ~~ 114 PBytes/s.
     // This speed is inaccessible to an ordinary user.
-    m_accumulator.x += pointData.x;
+    ++m_counter;
     for (int id = UP; id < NB_GRAPHS; ++id)
-        m_accumulator.y[id] += pointData.y[id];
-    m_counter = (m_counter + 1) % m_divider;
-    if (m_counter != 0)
-        return; // still accumulating
+        m_accumulator[id] += pointData[id];
+
+    const milliseconds elapsed {m_timeSinceLastPush.elapsed()};
+    if (elapsed < m_resolution)
+        return false; // still accumulating
+
     // it is time final averaging calculations
+    m_currentDuration += elapsed;
+    if (m_currentDuration > m_maxDuration)
+    {
+        m_sink.set_capacity(m_sink.size());
+        if (!m_sink.empty()) // front() will be removed on the following push so update the duration
+            m_currentDuration -= m_sink.front().duration;
+    }
+    else if (m_sink.size() + 1 >= m_sink.capacity())
+    {
+        m_sink.set_capacity(m_sink.size() + ((m_currentMaxDuration - m_currentDuration) / elapsed));
+    }
+
     for (int id = UP; id < NB_GRAPHS; ++id)
-        m_accumulator.y[id] /= m_divider;
-    m_accumulator.x /= m_divider;
+        m_accumulator[id] /= m_counter;
+
     // now flush out averaged data
-    m_sink.push_back(m_accumulator);
+    m_sink.push_back({elapsed, m_accumulator});
+
+    // reset
     m_accumulator = {};
+    m_counter = 0;
+    m_timeSinceLastPush.restart();
+    return true;
 }
 
-bool SpeedPlotView::Averager::isReady() const
+qint64 SpeedPlotView::Averager::viewablePointsCount() const
```

Comment:
Can you explain in simple words what `viewablePointsCount` is supposed to mean? What exactly does `viewable` mean in it? And why is it that the Averager, which appears to be a computational and accumulative helper, is responsible for something related to `viewing`?

Reply:
This was used to guess the full size of the buffer when max duration is reached, on the latest push I've made the SpeedPlotView fully duration based, removing the need for this function.